contracts/Curation/GardenTokenLock.sol

```diff
@@ -0,0 +1,66 @@
+// SPDX-License-Identifier: GPL-3.0
+
+pragma solidity =0.8.6;
+
+import "openzeppelin-contracts-upgradable-v4/access/OwnableUpgradeable.sol";
+import "openzeppelin-contracts-upgradable-v4/utils/math/SafeMathUpgradeable.sol";
+import "../Distributors/TokenManagerHook.sol";
+import "../Interfaces/ITokenManager.sol";
+
+contract GardenTokenLock is TokenManagerHook {
+    using SafeMathUpgradeable for uint256;
+
+    uint256 public initialDate;
+    uint256 public roundDuration;
+    IERC20 public token;
+
+    struct Lock {
+        uint256 totalAmountLocked;
+        mapping(uint256 => uint256) amountLockedUntilRound;
+    }
+
+    mapping(address => Lock) public lockedTokens;
+
+    error TokensAreLocked();
+    error CannotUnlockUntilRoundIsFinished();
+    
+    function __GardenTokenLock_init(uint256 _initialDate, uint256 _roundDuration, address _tokenManager) public initializer {
+        __TokenManagerHook_initialize(_tokenManager);
+        initialDate = _initialDate;
+        roundDuration = _roundDuration;
+        token = ITokenManager(_tokenManager).token();
+    }
+
+    function lock(uint256 _amount, uint256 _rounds) public virtual {
+        Lock storage _lock = lockedTokens[msg.sender];
```

Comment:
Check added, it was a virtual function, but it's better to have the check here than in the child function.

I think it's better to keep using the SafeMath library as it is used in the rest of the contracts, for consistency and also just in case we downgrade the solidity version in the future for some reason.